File: Common/Source/Android/Air3/PowerManagerUtils.cpp
//
// Created by BLC on 14/10/2022.
//

#include "PowerManagerUtils.h"
#include "Android/Context.hpp"
#include "Android/Main.hpp"
#include "Java/Class.hxx"

namespace {

Java::Class GetClass(JNIEnv * env) {
    Java::Class cls = { env, "org/LK8000/Air3/PowerManagerUtils"};
    Java::RethrowException(env); // rethrow ClassNotFoundException
    return cls;
}

} // namespace

bool PowerManagerUtils::openModuleFanet() noexcept {
    JNIEnv * env = Java::GetEnv();
    try {
        Java::Class cls = GetClass(env);

        jmethodID method = env->GetStaticMethodID(cls.Get(),
                                                  "openModuleFanet",
                                                  "(Landroid/content/Context;)Z");
        Java::RethrowException(env); // rethrow NoSuchMethodError
        if (!method) {
            return false;
        }
        env->CallStaticBooleanMethod(cls, method, context->Get());
    } catch(std::exception& e) {
        return false;
    }
    return true;
}

void PowerManagerUtils::closeModuleFanet() noexcept {
    JNIEnv * env = Java::GetEnv();
    try {
        Java::Class cls = GetClass(env);
        jmethodID method = env->GetStaticMethodID(cls.Get(),
                                                  "closeModuleFanet",
                                                  "(Landroid/content/Context;)V");
        Java::RethrowException(env); // rethrow NoSuchMethodError
        if (method) {
            env->CallStaticVoidMethod(cls, method, context->Get());
        }
    }
    catch (...) {}
}

File: android/src/org/LK8000/Air3/PowerManagerUtils.java
package org.LK8000.Air3;

import android.content.Context;
import android.os.PowerManager;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Objects;

public class PowerManagerUtils {
    private static final String CLASS = "android.os.PowerManager";
    private static final String UART_POWER_OPEN = "modulePower";

    public static final Integer FANET_MODULE = 2;

    private static PowerManager getPowerManager(Context context) throws NullPointerException {
        return Objects.requireNonNull((PowerManager) context.getSystemService(Context.POWER_SERVICE));
    }

    private static void invoke(PowerManager powerManager, Integer i, Boolean b) throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, IllegalAccessException {
        Method method = Class.forName(CLASS).getMethod(UART_POWER_OPEN, Integer.TYPE, Boolean.TYPE);
        method.invoke(powerManager, i, b);
    }

    private static void close(PowerManager powerManager, Integer i) {
        try {
            invoke(powerManager, i, Boolean.FALSE);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static boolean open(PowerManager powerManager, Integer i) {
        try {
            invoke(powerManager, i, Boolean.TRUE);
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    public static boolean openModuleFanet(Context context) {
        PowerManager pm = PowerManagerUtils.getPowerManager(context);
        return PowerManagerUtils.open(pm, PowerManagerUtils.FANET_MODULE);
    }

    public static void closeModuleFanet(Context context) {
        PowerManager pm = PowerManagerUtils.getPowerManager(context);
        PowerManagerUtils.close(pm, PowerManagerUtils.FANET_MODULE);
    }
}
